validator.go

```diff
@@ -101,3 +104,131 @@ func newValidatorImpl(domains []string, usersFile string,
+		return len(v) > 0
+	case map[string]interface{}:
+		return len(v) > 0
+	case string:
+		return len(v) > 0
+	case nil:
+		return false
+	}
+
+	// go-jmespath does extra validation as well, we should keep parity
+	rv := reflect.ValueOf(result)
+	switch rv.Kind() {
+	case reflect.Struct:
+		// Structs are not the same as an empty map (i.e. they are "something"
+		// even if all 0's of something), thus true here.
+		return true
+	case reflect.Slice, reflect.Map:
+		return rv.Len() > 0
+	case reflect.Ptr:
+		if rv.IsNil() {
+			return false
+		}
+		// If a pointer, check the pointed at value.
+		elem := rv.Elem()
+		return truthy(elem.Interface())
+	}
+
+	return true
+}
+
+func (v *JMESValidator) AddRule(jmespathExpr string) (bool, error) {
+
+	// TODO: Check for duplicate rules?
+
+	rule := strings.TrimSpace(jmespathExpr)
+	if rule != "" && !strings.HasPrefix(rule, "#") {
+		var compiled *jmespath.JMESPath
+		var err error
+
+		if compiled, err = jmespath.Compile(rule); err != nil {
+			return false, fmt.Errorf("invalid claim assertion (%q): %v", rule, err)
+		}
+
+		// Invalidate the hash if it had been requested yet
+		v.rulesHash = nil
+		v.rules = append(v.rules, rule)
+		v.compiledRules = append(v.compiledRules, compiled)
+
+		return true, nil
+	}
+
+	// Not an error, but nothing as added either (empty or commented rule)
+	return false, nil
+}
+
+// Rules will return the current set of valid registered rules (in source form).
+// Edits to these rules do not affect the internal validation as they have already
+// been pre-compiled by the time they are added to this list.
+func (v *JMESValidator) Rules() []string {
+	return v.rules
+}
+
+// RulesHash returns a deterministic hashed value of the input rules that have been
+// registered in this validator. It's meant to serve as a quick way to know if the
+// set of rules has changed over time.
+func (v *JMESValidator) RulesHash() []byte {
+
+	if v.rulesHash == nil && len(v.rules) > 0 {
+
+		// Create a hash of our rules (in source form) so that we can know
+		// if they have changed since they were last run against something
+		h := crypto.SHA256.New()
+		for _, rule := range v.rules {
+			h.Write([]byte(rule))
+		}
+
+		v.rulesHash = h.Sum(nil)
+	}
+
+	return v.rulesHash
+}
+
+// MatchesAny will scan the list of compiled rules, in order, and return true
+// on the first match. If the provided data parameter is nil, or no rules
+// match, return false. Note: an empty set of rules does NOT return true,
+// since at least one rule must match. If a match is found, returns true as
+// well as the index of the rule that was successfully matched.
+func (v *JMESValidator) MatchesAny(data map[string]interface{}) (bool, int) {
+
+	// if s, err := json.MarshalIndent(data, "", "  "); err == nil {
+	// 	logger.Printf("CLAIMS: %s", string(s))
+	// }
+
+	if data == nil {
+		return false, -1
+	}
+
+	for idx, rule := range v.compiledRules {
+		if result, err := rule.Search(data); err == nil {
```

Comment:
What kind of errors are returned from this? Should we be doing something about them?

Reply:
If the data isn't what the expression is expecting, (like maybe taking the length of a boolean?), things like that. About the only thing you could do is log that a specific rule failed, but it also may be intentional if your provider gives strange otherwise different results in some cases in which case it's erroneous log spam for those. It also isn't great to log some of these claims' values in the case of them being personally identifiable information and you care about that.